java/vector/src/main/java/org/apache/arrow/vector/file/json/JsonFileReader.java

```diff
@@ -234,29 +229,42 @@ private void readVector(Field field, FieldVector vector) throws JsonParseExcepti
       if (started && !Objects.equal(field.getName(), name)) {
         throw new IllegalArgumentException("Expected field " + field.getName() + " but got " + name);
       }
+
+      // Initialize the vector with required capacity
       int count = readNextField("count", Integer.class);
+      vector.setInitialCapacity(count);
       vector.allocateNew();
-      vector.getMutator().setValueCount(count);
+
+      // Read inner vectors
       for (int v = 0; v < vectorTypes.size(); v++) {
         ArrowVectorType vectorType = vectorTypes.get(v);
-        BufferBacked innerVector = fieldInnerVectors.get(v);
+        ValueVector valueVector = (ValueVector) fieldInnerVectors.get(v);
         nextFieldIs(vectorType.getName());
         readToken(START_ARRAY);
-        ValueVector valueVector = (ValueVector) innerVector;
-
         int innerVectorCount = vectorType.equals(OFFSET) ? count + 1 : count;
-        valueVector.setInitialCapacity(innerVectorCount);
-        valueVector.allocateNew();
-
         for (int i = 0; i < innerVectorCount; i++) {
           parser.nextToken();
           setValueFromParser(valueVector, i);
         }
-        Mutator mutator = valueVector.getMutator();
-        mutator.setValueCount(innerVectorCount);
         readToken(END_ARRAY);
       }
-      // if children
+
+      // Set lastSet before valueCount to prevent setValueCount from filling empty values
+      switch (vector.getMinorType()) {
```

Comment:
I am not crazy about this because case matching makes it hard to maintain for new types of Vectors.

I think the proper way to do this is to use `loadFieldBuffers` in this class, since loadFieldBuffers should initialize vectors correctly. I think ArrowFileReader uses it too.

Reply:
Yeah, it's not ideal in that sense but `loadFieldBuffers` uses nodes that come from FB data in `ArrowRecordBatch`s. I think it was designed only for that type of data, so I'm not sure how to apply  it here for json.